File: tests/test_api_key_save.py
"""
Test script to verify API key saving functionality
Tests encryption and database operations
"""
import os
import sys
import subprocess
import json

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

def test_encryption_key():
    """Test if ENCRYPTION_KEY is set and valid"""
    print("Testing ENCRYPTION_KEY...")
    encryption_key = os.getenv("ENCRYPTION_KEY")
    
    if not encryption_key:
        print("❌ ENCRYPTION_KEY not set in environment")
        return False
    
    if len(encryption_key) < 32:
        print(f"❌ ENCRYPTION_KEY too short: {len(encryption_key)} chars (need 32+)")
        return False
    
    print(f"✅ ENCRYPTION_KEY is set ({len(encryption_key)} chars)")
    return True

def test_database_connection():
    """Test database connection"""
    print("\nTesting database connection...")
    database_url = os.getenv("DATABASE_URL")
    
    if not database_url:
        print("❌ DATABASE_URL not set")
        return False
    
    print(f"✅ DATABASE_URL is set")
    
    # Try to connect (requires psycopg2 or similar)
    try:
        import psycopg2
        from urllib.parse import urlparse
        
        parsed = urlparse(database_url)
        conn = psycopg2.connect(
            host=parsed.hostname,
            port=parsed.port or 5432,
            database=parsed.path[1:],  # Remove leading /
            user=parsed.username,
            password=parsed.password
        )
        
        # Check if api_keys table exists
        cursor = conn.cursor()
        cursor.execute("""
            SELECT EXISTS (
                SELECT FROM information_schema.tables 
                WHERE table_name = 'api_keys'
            );
        """)
        exists = cursor.fetchone()[0]
        
        if exists:
            print("✅ api_keys table exists")
            
            # Check table structure
            cursor.execute("""
                SELECT column_name, data_type 
                FROM information_schema.columns 
                WHERE table_name = 'api_keys'
                ORDER BY ordinal_position;
            """)
            columns = cursor.fetchall()
            print(f"   Table has {len(columns)} columns:")
            for col_name, col_type in columns:
                print(f"   - {col_name}: {col_type}")
        else:
            print("❌ api_keys table does not exist - run migrations!")
            cursor.close()
            conn.close()
            return False
        
        cursor.close()
        conn.close()
        return True
        
    except ImportError:
        print("⚠️  psycopg2 not installed - skipping database connection test")
        print("   Install with: pip install psycopg2-binary")
        return None
    except Exception as e:
        print(f"❌ Database connection failed: {e}")
        return False

def main():
    """Run all tests"""
    print("=" * 60)
    print("API Key Save Test")
    print("=" * 60)
    
    # Load .env.local if it exists
    env_file = os.path.join(os.path.dirname(os.path.dirname(__file__)), ".env.local")
    if os.path.exists(env_file):
        print(f"\nLoading environment from {env_file}...")
        with open(env_file, 'r') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#') and '=' in line:
                    key, value = line.split('=', 1)
                    os.environ[key.strip()] = value.strip()
    
    results = []
    
    # Test encryption key
    results.append(("Encryption Key", test_encryption_key()))
    
    # Test database
    db_result = test_database_connection()
    if db_result is not None:
        results.append(("Database Connection", db_result))
    
    # Summary
    print("\n" + "=" * 60)
    print("Test Summary")
    print("=" * 60)
    for test_name, result in results:
        status = "✅ PASS" if result else "❌ FAIL"
        print(f"{test_name}: {status}")
    
    all_passed = all(result for _, result in results if result is not None)
    
    if all_passed:
        print("\n✅ All tests passed!")
        return 0
    else:
        print("\n❌ Some tests failed. Check the output above.")
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: tests/test-api-key-save.js
/**
 * Test script to verify API key saving functionality
 * Tests encryption and database operations
 */

const { Pool } = require('pg');
require('dotenv').config({ path: '.env.local' });

async function testEncryptionKey() {
    console.log('\n📝 Testing ENCRYPTION_KEY...');
    const encryptionKey = process.env.ENCRYPTION_KEY;
    
    if (!encryptionKey) {
        console.error('❌ ENCRYPTION_KEY not set in environment');
        return false;
    }
    
    if (encryptionKey.length < 32) {
        console.error(`❌ ENCRYPTION_KEY too short: ${encryptionKey.length} chars (need 32+)`);
        return false;
    }
    
    console.log(`✅ ENCRYPTION_KEY is set (${encryptionKey.length} chars)`);
    return true;
}

async function testDatabaseConnection() {
    console.log('\n📝 Testing database connection...');
    const databaseUrl = process.env.DATABASE_URL;
    
    if (!databaseUrl) {
        console.error('❌ DATABASE_URL not set');
        return false;
    }
    
    console.log('✅ DATABASE_URL is set');
    
    const pool = new Pool({
        connectionString: databaseUrl,
    });
    
    try {
        // Test connection
        await pool.query('SELECT 1');
        console.log('✅ Database connection successful');
        
        // Check if api_keys table exists
        const tableCheck = await pool.query(`
            SELECT EXISTS (
                SELECT FROM information_schema.tables 
                WHERE table_name = 'api_keys'
            );
        `);
        
        if (!tableCheck.rows[0].exists) {
            console.error('❌ api_keys table does not exist - run migrations!');
            console.log('   Run: npm run db:push or npm run db:migrate');
            await pool.end();
            return false;
        }
        
        console.log('✅ api_keys table exists');
        
        // Check table structure
        const columns = await pool.query(`
            SELECT column_name, data_type, is_nullable
            FROM information_schema.columns 
            WHERE table_name = 'api_keys'
            ORDER BY ordinal_position;
        `);
        
        console.log(`   Table has ${columns.rows.length} columns:`);
        columns.rows.forEach(col => {
            console.log(`   - ${col.column_name}: ${col.data_type} (nullable: ${col.is_nullable})`);
        });
        
        await pool.end();
        return true;
        
    } catch (error) {
        console.error('❌ Database connection failed:', error.message);
        await pool.end();
        return false;
    }
}

async function testEncryption() {
    console.log('\n📝 Testing encryption functions...');
    
    try {
        // Try to import and use the encryption functions
        const crypto = require('../src/lib/crypto.ts');
        
        const testKey = 'sk-test123456789012345678901234567890';
        const encrypted = crypto.encryptApiKey(testKey);
        console.log('✅ Encryption successful');
        console.log(`   Encrypted length: ${encrypted.length} chars`);
        
        const decrypted = crypto.decryptApiKey(encrypted);
        if (decrypted === testKey) {
            console.log('✅ Decryption successful - roundtrip works');
            return true;
        } else {
            console.error('❌ Decryption failed - keys do not match');
            return false;
        }
        
    } catch (error) {
        console.error('❌ Encryption test failed:', error.message);
        console.error('   Stack:', error.stack);
        return false;
    }
}

async function main() {
    console.log('='.repeat(60));
    console.log('API Key Save Test');
    console.log('='.repeat(60));
    
    const results = [];
    
    // Test encryption key
    results.push(['Encryption Key', await testEncryptionKey()]);
    
    // Test database
    results.push(['Database Connection', await testDatabaseConnection()]);
    
    // Test encryption functions (if possible)
    try {
        results.push(['Encryption Functions', await testEncryption()]);
    } catch (error) {
        console.log('\n⚠️  Skipping encryption function test (TypeScript module)');
    }
    
    // Summary
    console.log('\n' + '='.repeat(60));
    console.log('Test Summary');
    console.log('='.repeat(60));
    results.forEach(([name, result]) => {
        const status = result ? '✅ PASS' : '❌ FAIL';
        console.log(`${name}: ${status}`);
    });
    
    const allPassed = results.every(([, result]) => result);
    
    if (allPassed) {
        console.log('\n✅ All tests passed!');
        process.exit(0);
    } else {
        console.log('\n❌ Some tests failed. Check the output above.');
        process.exit(1);
    }
}

main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
});
